## Task -01 (ELA AND BUSINESS INSIGHTS)

In [62]:
import pandas as pd

# Load the datasets
customers_df = pd.read_csv("Customers.csv")
products_df = pd.read_csv("Products.csv")
transactions_df = pd.read_csv("Transactions.csv")

# Display the first 5 rows of each DataFrame
print("Customers DataFrame:")
print(customers_df.head().to_markdown(index=False, numalign="left", stralign="left"))

print("\nProducts DataFrame:")
print(products_df.head().to_markdown(index=False, numalign="left", stralign="left"))

print("\nTransactions DataFrame:")
print(transactions_df.head().to_markdown(index=False, numalign="left", stralign="left"))

# Display columns and their data types for each DataFrame
print("\nCustomers DataFrame Info:")
print(customers_df.info())

print("\nProducts DataFrame Info:")
print(products_df.info())

print("\nTransactions DataFrame Info:")
print(transactions_df.info())

Customers DataFrame:
| CustomerID   | CustomerName       | Region        | SignupDate   |
|:-------------|:-------------------|:--------------|:-------------|
| C0001        | Lawrence Carroll   | South America | 2022-07-10   |
| C0002        | Elizabeth Lutz     | Asia          | 2022-02-13   |
| C0003        | Michael Rivera     | South America | 2024-03-07   |
| C0004        | Kathleen Rodriguez | South America | 2022-10-09   |
| C0005        | Laura Weber        | Asia          | 2022-08-15   |

Products DataFrame:
| ProductID   | ProductName             | Category    | Price   |
|:------------|:------------------------|:------------|:--------|
| P001        | ActiveWear Biography    | Books       | 169.3   |
| P002        | ActiveWear Smartwatch   | Electronics | 346.3   |
| P003        | ComfortLiving Biography | Books       | 44.12   |
| P004        | BookWorld Rug           | Home Decor  | 95.69   |
| P005        | TechPro T-Shirt         | Clothing    | 429.31  |

Transactions

##  Checking the missing values.

In [65]:
# Check for missing values in each DataFrame
print("Missing values in Customers DataFrame:")
print(customers_df.isnull().sum())

print("\nMissing values in Products DataFrame:")
print(products_df.isnull().sum())

print("\nMissing values in Transactions DataFrame:")
print(transactions_df.isnull().sum())

Missing values in Customers DataFrame:
CustomerID      0
CustomerName    0
Region          0
SignupDate      0
dtype: int64

Missing values in Products DataFrame:
ProductID      0
ProductName    0
Category       0
Price          0
dtype: int64

Missing values in Transactions DataFrame:
TransactionID      0
CustomerID         0
ProductID          0
TransactionDate    0
Quantity           0
TotalValue         0
Price              0
dtype: int64


## TransactionDate column in the transactions_df DataFrame to datetime format

In [68]:
# Convert TransactionDate to datetime
transactions_df['TransactionDate'] = pd.to_datetime(transactions_df['TransactionDate'])

# Display the first 5 rows of the updated transactions_df
print("Updated Transactions DataFrame:")
print(transactions_df.head().to_markdown(index=False, numalign="left", stralign="left"))

# Display columns and their data types to check if the conversion was successful
print("\nUpdated Transactions DataFrame Info:")
print(transactions_df.info())

Updated Transactions DataFrame:
| TransactionID   | CustomerID   | ProductID   | TransactionDate     | Quantity   | TotalValue   | Price   |
|:----------------|:-------------|:------------|:--------------------|:-----------|:-------------|:--------|
| T00001          | C0199        | P067        | 2024-08-25 12:38:23 | 1          | 300.68       | 300.68  |
| T00112          | C0146        | P067        | 2024-05-27 22:23:54 | 1          | 300.68       | 300.68  |
| T00166          | C0127        | P067        | 2024-04-25 07:38:55 | 1          | 300.68       | 300.68  |
| T00272          | C0087        | P067        | 2024-03-26 22:55:37 | 2          | 601.36       | 300.68  |
| T00363          | C0070        | P067        | 2024-03-21 15:10:10 | 3          | 902.04       | 300.68  |

Updated Transactions DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype         
---  ----

### I'll start by analyzing the Customers dataset

In [72]:
# Calculate the number of unique values in CustomerID, CustomerName, and Region
unique_customer_ids = customers_df['CustomerID'].nunique()
unique_customer_names = customers_df['CustomerName'].nunique()
unique_regions = customers_df['Region'].nunique()

# Display the results
print("Number of unique Customer IDs:", unique_customer_ids)
print("Number of unique Customer Names:", unique_customer_names)
print("Number of unique Regions:", unique_regions)

Number of unique Customer IDs: 200
Number of unique Customer Names: 200
Number of unique Regions: 4


In [74]:
### Now Products Dataset

In [76]:
# Calculate the number of unique values in ProductID, ProductName, and Category
unique_product_ids = products_df['ProductID'].nunique()
unique_product_names = products_df['ProductName'].nunique()
unique_categories = products_df['Category'].nunique()

# Display the results
print("Number of unique Product IDs:", unique_product_ids)
print("Number of unique Product Names:", unique_product_names)
print("Number of unique Categories:", unique_categories)

Number of unique Product IDs: 100
Number of unique Product Names: 66
Number of unique Categories: 4


In [78]:
# Calculate and display the 10 most frequently occurring products
top_10_products = transactions_df['ProductID'].value_counts().head(10)
print("Top 10 most frequently occurring products:\n")
print(top_10_products.to_markdown(numalign="left", stralign="left"))

Top 10 most frequently occurring products:

| ProductID   | count   |
|:------------|:--------|
| P059        | 19      |
| P029        | 17      |
| P062        | 16      |
| P079        | 16      |
| P054        | 16      |
| P061        | 16      |
| P048        | 15      |
| P022        | 15      |
| P096        | 15      |
| P049        | 15      |


In [80]:
# Merge products_df and transactions_df on ProductID
merged_df = pd.merge(products_df, transactions_df, on='ProductID')

# Calculate and display the top 10 product categories
top_10_categories = merged_df['Category'].value_counts().head(10)
print("Top 10 most frequently bought product categories:\n")
print(top_10_categories.to_markdown(numalign="left", stralign="left"))

Top 10 most frequently bought product categories:

| Category    | count   |
|:------------|:--------|
| Books       | 270     |
| Electronics | 254     |
| Home Decor  | 248     |
| Clothing    | 228     |


In [82]:
# Display descriptive statistics of Quantity and TotalValue
print("Descriptive statistics of Quantity and TotalValue:\n")
print(transactions_df[['Quantity', 'TotalValue']].describe().to_markdown(numalign="left", stralign="left"))

Descriptive statistics of Quantity and TotalValue:

|       | Quantity   | TotalValue   |
|:------|:-----------|:-------------|
| count | 1000       | 1000         |
| mean  | 2.537      | 689.996      |
| std   | 1.11798    | 493.144      |
| min   | 1          | 16.08        |
| 25%   | 2          | 295.295      |
| 50%   | 3          | 588.88       |
| 75%   | 4          | 1011.66      |
| max   | 4          | 1991.04      |


### On average, customers purchase 2 to 3 items per transaction, with an average total value of 689.99. The maximum TotalValue is 1991.04, significantly higher than the mean, suggesting potential outliers.

In [88]:
# Merge customers_df and transactions_df on CustomerID
customer_transactions_df = pd.merge(customers_df, transactions_df, on='CustomerID')

# Group by Region and count the number of transactions
transactions_by_region = customer_transactions_df.groupby('Region').size().reset_index(name='TransactionCount')

# Display the first 5 rows of the grouped DataFrame
print("Transactions by Region:")
print(transactions_by_region.head().to_markdown(index=False, numalign="left", stralign="left"))

Transactions by Region:
| Region        | TransactionCount   |
|:--------------|:-------------------|
| Asia          | 218                |
| Europe        | 234                |
| North America | 244                |
| South America | 304                |


In [90]:
import altair as alt

# Create the bar chart
chart = alt.Chart(transactions_by_region).mark_bar().encode(
    x=alt.X('Region'),
    y=alt.Y('TransactionCount'),
    tooltip = ['Region', 'TransactionCount']
).properties(
    title='Distribution of Transactions by Region'
).interactive()

# Save the chart
chart.save('transaction_distribution_by_region_bar_chart.json')

C:\Users\chand\anaconda3\Lib\site-packages\altair\utils\core.py:395: FutureWarning: the convert_dtype parameter is deprecated and will be removed in a future version.  Do ``ser.astype(object).apply()`` instead if you want ``convert_dtype=False``.
  col = df[col_name].apply(to_list_if_array, convert_dtype=False)


### >>>South America has the highest number of transactions (304), followed by North America (244).